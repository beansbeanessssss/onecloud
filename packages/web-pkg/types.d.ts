import { Ability } from 'web-client/src/helpers/resource/types'
import { ClientService, LoadingService, PreviewService, ArchiverService } from './src/services'

export * from './src'

declare module 'vue' {
  interface ComponentCustomProperties {
    $ability: Ability
    $archiverService: ArchiverService
    $clientService: ClientService
    $loadingService: LoadingService
    $previewService: PreviewService
  }
}
