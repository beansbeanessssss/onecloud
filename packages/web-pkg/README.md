# web-pkg

This is experimental.
