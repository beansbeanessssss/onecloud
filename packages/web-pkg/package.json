{
  "name": "@ownclouders/web-pkg",
  "version": "0.0.0",
  "description": "ownCloud web pkg",
  "license": "AGPL-3.0",
  "main": "src/index.ts",
  "peerDependencies": {
    "axios": "^0.27.2",
    "filesize": "^9.0.11",
    "fuse.js": "^6.5.3",
    "lodash-es": "^4.17.21",
    "luxon": "^2.4.0",
    "mark.js": "^8.11.1",
    "popper-max-size-modifier": "^0.2.0",
    "qs": "^6.10.3",
    "uuid": "^9.0.0",
    "vue-router": "4.1.6",
    "vuex": "4.1.0",
    "web-client": "npm:@ownclouders/web-client",
    "web-pkg": "npm:@ownclouders/web-pkg",
    "vue-concurrency": "4.0.0"
  }
}
