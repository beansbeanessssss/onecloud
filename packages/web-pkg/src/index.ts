export * from './components'
export * from './composables'
export * from './configuration'
export * from './errors'
export * from './helpers'
export * from './http'
export * from './observer'
export * from './services'
export * from './utils'
