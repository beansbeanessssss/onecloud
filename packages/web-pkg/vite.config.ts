import { join, resolve } from 'path'
import { defineConfig, searchForWorkspaceRoot } from 'vite'
import dts from 'vite-plugin-dts'
import { nodePolyfills } from 'vite-plugin-node-polyfills'
import vue from '@vitejs/plugin-vue'

const projectRootDir = searchForWorkspaceRoot(process.cwd())

export default defineConfig({
  resolve: {
    alias: {
      crypto: join(projectRootDir, 'polyfills/crypto.js')
    }
  },
  css: {
    preprocessorOptions: {
      scss: {
        additionalData: `@import "design-system/src/styles/styles";`
      }
    }
  },
  build: {
    lib: {
      entry: resolve(__dirname, 'src/index.ts'),
      name: 'web-pkg',
      fileName: 'web-pkg'
    }
  },
  plugins: [
    vue(),
    nodePolyfills({
      exclude: ['crypto']
    }),
    dts({ declarationOnly: true, exclude: ['**/tests'] }),
    {
      name: '@ownclouders/vite-plugin-docs',
      transform(src, id) {
        if (id.includes('type=docs')) {
          return {
            code: 'export default {}',
            map: null
          }
        }
      }
    }
  ]
})
