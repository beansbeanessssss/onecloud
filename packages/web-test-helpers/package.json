{
  "name": "web-test-helpers",
  "private": true,
  "main": "src/index.ts",
  "peerDependencies": {
    "axios": "0.27.2",
    "vue-gettext": "2.1.12",
    "vue-router": "3.6.5",
    "vuex": "4.1.0"
  }
}
