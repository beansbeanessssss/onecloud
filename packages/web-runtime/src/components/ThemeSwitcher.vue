<template>
  <div>
    <oc-select
      :model-value="currentTheme"
      :clearable="false"
      :options="availableThemes"
      option-label="name"
      @update:model-value="updateTheme"
    />
  </div>
</template>
<script lang="ts">
import { defineComponent, unref } from 'vue'
import { useMessages, useThemeStore, WebThemeType } from '@ownclouders/web-pkg'
import { storeToRefs } from 'pinia'
import { useGettext } from 'vue3-gettext'

export default defineComponent({
  setup() {
    const themeStore = useThemeStore()
    const { showMessage } = useMessages()
    const { $gettext } = useGettext()

    const { setAndApplyTheme } = themeStore

    const updateTheme = (newTheme: WebThemeType) => {
      setAndApplyTheme(newTheme)
      showMessage({ title: $gettext('Preference saved.') })
    }

    const { availableThemes, currentTheme } = storeToRefs(themeStore)

    const isThemeCurrentlyActive = (theme: WebThemeType) => {
      return unref(currentTheme) === theme
    }

    return {
      availableThemes,
      currentTheme,
      isThemeCurrentlyActive,
      updateTheme
    }
  }
})
</script>
