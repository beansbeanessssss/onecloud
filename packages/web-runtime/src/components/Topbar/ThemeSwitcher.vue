<template>
  <div v-if="!hasOnlyOneTheme">
    <oc-button
      v-if="hasOnlyTwoThemesForLightDarkMode"
      v-oc-tooltip="buttonLabel"
      class="themeswitcher-btn"
      :aria-label="buttonLabel"
      appearance="raw-inverse"
      variation="brand"
      @click="toggleTheme"
    >
      <span class="oc-visible@s" :aria-label="switchLabel" />
      <oc-icon :name="switchIcon" fill-type="line" variation="inherit" />
    </oc-button>
    <template v-else>
      <oc-button
        id="theme-picker-options-btn"
        key="theme-picker-options-btn"
        v-oc-tooltip="buttonLabel"
        data-testid="theme-picker-options-btn"
        :aria-label="buttonLabel"
        appearance="raw-inverse"
        variation="brand"
      >
        <oc-icon name="palette" />
      </oc-button>
      <oc-drop
        drop-id="theme-picker-options-drop"
        toggle="#theme-picker-options-btn"
        mode="click"
        padding-size="small"
        close-on-click
      >
        <oc-list class="oc-tiles-sort-list">
          <li v-for="(theme, index) in availableThemes" :key="index" class="oc-my-xs">
            <oc-button
              :aria-label="buttonLabel"
              justify-content="space-between"
              class="theme-picker-options-list-item oc-p-s oc-width-1-1"
              :class="{
                'oc-background-primary-gradient': isThemeCurrentlyActive(theme),
                selected: isThemeCurrentlyActive(theme)
              }"
              :appearance="isThemeCurrentlyActive(theme) ? 'raw-inverse' : 'raw'"
              :variation="isThemeCurrentlyActive(theme) ? 'primary' : 'passive'"
              @click="setAndApplyTheme(theme)"
            >
              <span>{{ theme.name }}</span>
              <oc-icon v-if="isThemeCurrentlyActive(theme)" name="check" variation="inherit" />
            </oc-button>
          </li>
        </oc-list>
      </oc-drop>
    </template>
  </div>
</template>
<script lang="ts">
import { computed, defineComponent, unref } from 'vue'
import { useGettext } from 'vue3-gettext'
import { useThemeStore } from '@ownclouders/web-pkg'
import { storeToRefs } from 'pinia'

export default defineComponent({
  setup() {
    const { $gettext } = useGettext()
    const themeStore = useThemeStore()

    const { setAndApplyTheme, toggleTheme } = themeStore

    const { availableThemes, currentTheme, hasOnlyOneTheme, hasOnlyTwoThemesForLightDarkMode } =
      storeToRefs(themeStore)

    const buttonLabel = computed(() => $gettext('Click to switch theme'))
    const switchLabel = computed(() => $gettext('Currently used theme'))
    const switchIcon = computed(() => (currentTheme.value.isDark === false ? 'sun' : 'moon-clear'))

    const isThemeCurrentlyActive = (theme) => {
      return unref(currentTheme) === theme
    }

    return {
      availableThemes,
      buttonLabel,
      currentTheme,
      hasOnlyOneTheme,
      hasOnlyTwoThemesForLightDarkMode,
      isThemeCurrentlyActive,
      setAndApplyTheme,
      switchIcon,
      switchLabel,
      toggleTheme
    }
  }
})
</script>
