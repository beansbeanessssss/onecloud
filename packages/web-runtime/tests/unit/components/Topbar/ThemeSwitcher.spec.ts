import { WebThemeType, useThemeStore } from '@ownclouders/web-pkg'
import { mock } from 'jest-mock-extended'
import ThemeSwitcher from 'web-runtime/src/components/Topbar/ThemeSwitcher.vue'
import defaultTheme from 'web-runtime/themes/owncloud/theme.json'
import { defaultPlugins, defaultStubs, mount } from 'web-test-helpers'

const defaultOwnCloudTheme = {
  defaults: {
    ...defaultTheme.clients.web.defaults,
    common: defaultTheme.common
  },
  themes: defaultTheme.clients.web.themes
}

describe('ThemeSwitcher component', () => {
  describe('visually', () => {
    beforeEach(() => {
      mockDarkModePreferred(false)
    })
    afterEach(() => {
      jest.clearAllMocks()
    })

    it('does not render anything if only one theme is available', async () => {
      const { wrapper } = getWrapper({ hasOnlyOneTheme: true })
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).toMatchSnapshot()
    })

    describe('if there is exactly two themes, one dark and one light', () => {
      it('renders a button, initially in light mode per default', async () => {
        const { wrapper } = getWrapper()
        await wrapper.vm.$nextTick()
        expect(wrapper.html()).toMatchSnapshot()
      })

      it('renders a button, initially in dark mode if user prefers dark color scheme', async () => {
        mockDarkModePreferred(true)
        const { wrapper } = getWrapper()
        await wrapper.vm.$nextTick()
        expect(wrapper.html()).toMatchSnapshot()
      })

      it('toggles between themes upon click', async () => {
        const { wrapper } = getWrapper()
        await wrapper.find('.themeswitcher-btn').trigger('click')
        await wrapper.vm.$nextTick()
        expect(wrapper.html()).toMatchSnapshot()
        await wrapper.find('.themeswitcher-btn').trigger('click')
        await wrapper.vm.$nextTick()
        expect(wrapper.html()).toMatchSnapshot()
      })
    })

    describe('if there are more than two themes', () => {
      it.todo('renders a button, with a dropdown menu')
    })
  })

  describe('restores', () => {
    it('light theme if light theme is saved in localstorage', async () => {
      const themeStore = useThemeStore()
      window.localStorage.setItem('oc_currentThemeName', 'Light Theme')
      const { wrapper } = getWrapper({ hasOnlyOneTheme: false })
      themeStore.initializeThemes(defaultOwnCloudTheme)
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('dark theme if dark theme is saved in localstorage', async () => {
      const themeStore = useThemeStore()
      window.localStorage.setItem('oc_currentThemeName', 'Dark Theme')
      const { wrapper } = getWrapper()
      themeStore.initializeThemes(defaultOwnCloudTheme)
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})

function mockDarkModePreferred(enabled = false) {
  // matchMedia is not implemented in JSDOM, needs to be mocked
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(() => ({
      matches: enabled,
      media: "'(prefers-color-scheme: dark)'",
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn()
    }))
  })
}

function getWrapper({ hasOnlyOneTheme = false } = {}) {
  const availableThemes = hasOnlyOneTheme
    ? [defaultTheme.clients.web.themes[0]]
    : defaultTheme.clients.web.themes

  return {
    wrapper: mount(ThemeSwitcher, {
      global: {
        plugins: [
          ...defaultPlugins({
            piniaOptions: {
              stubActions: false,
              themeState: {
                availableThemes,
                currentTheme: mock<WebThemeType>({
                  ...defaultOwnCloudTheme.defaults,
                  ...defaultOwnCloudTheme.themes[0]
                })
              }
            }
          })
        ],
        stubs: { ...defaultStubs, 'oc-icon': true }
      }
    })
  }
}
