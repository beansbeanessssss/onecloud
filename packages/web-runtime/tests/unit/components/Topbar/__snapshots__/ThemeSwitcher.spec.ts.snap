// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitcher component restores dark theme if dark theme is saved in localstorage 1`] = `
<div>
  <button aria-label="Click to switch theme" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-brand oc-button-brand-raw-inverse themeswitcher-btn" type="button">
    <!--v-if-->
    <!-- @slot Content of the button -->
    <span aria-label="Currently used theme" class="oc-visible@s"></span>
    <oc-icon-stub fill-type="line" name="moon-clear" variation="inherit"></oc-icon-stub>
  </button>
</div>
`;

exports[`ThemeSwitcher component restores light theme if light theme is saved in localstorage 1`] = `
<div>
  <button aria-label="Click to switch theme" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-brand oc-button-brand-raw-inverse themeswitcher-btn" type="button">
    <!--v-if-->
    <!-- @slot Content of the button -->
    <span aria-label="Currently used theme" class="oc-visible@s"></span>
    <oc-icon-stub fill-type="line" name="sun" variation="inherit"></oc-icon-stub>
  </button>
</div>
`;

exports[`ThemeSwitcher component visually does not render anything if only one theme is available 1`] = `<!--v-if-->`;

exports[`ThemeSwitcher component visually if there is exactly two themes, one dark and one light renders a button, initially in dark mode if user prefers dark color scheme 1`] = `
<div>
  <button aria-label="Click to switch theme" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-brand oc-button-brand-raw-inverse themeswitcher-btn" type="button">
    <!--v-if-->
    <!-- @slot Content of the button -->
    <span aria-label="Currently used theme" class="oc-visible@s"></span>
    <oc-icon-stub fill-type="line" name="sun" variation="inherit"></oc-icon-stub>
  </button>
</div>
`;

exports[`ThemeSwitcher component visually if there is exactly two themes, one dark and one light renders a button, initially in light mode per default 1`] = `
<div>
  <button aria-label="Click to switch theme" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-brand oc-button-brand-raw-inverse themeswitcher-btn" type="button">
    <!--v-if-->
    <!-- @slot Content of the button -->
    <span aria-label="Currently used theme" class="oc-visible@s"></span>
    <oc-icon-stub fill-type="line" name="sun" variation="inherit"></oc-icon-stub>
  </button>
</div>
`;

exports[`ThemeSwitcher component visually if there is exactly two themes, one dark and one light toggles between themes upon click 1`] = `
<div>
  <button aria-label="Click to switch theme" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-brand oc-button-brand-raw-inverse themeswitcher-btn" type="button">
    <!--v-if-->
    <!-- @slot Content of the button -->
    <span aria-label="Currently used theme" class="oc-visible@s"></span>
    <oc-icon-stub fill-type="line" name="moon-clear" variation="inherit"></oc-icon-stub>
  </button>
</div>
`;

exports[`ThemeSwitcher component visually if there is exactly two themes, one dark and one light toggles between themes upon click 2`] = `
<div>
  <button aria-label="Click to switch theme" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-brand oc-button-brand-raw-inverse themeswitcher-btn" type="button">
    <!--v-if-->
    <!-- @slot Content of the button -->
    <span aria-label="Currently used theme" class="oc-visible@s"></span>
    <oc-icon-stub fill-type="line" name="sun" variation="inherit"></oc-icon-stub>
  </button>
</div>
`;
